### PR TITLE
[dtls] refine CoAP secure and DTLS

### PR DESCRIPTION
**Pull request by bukepo**
`CoapSecure` is always used by joiner, border agent or commissioner.
DTLS should not be shared with the application CoapSecure.

This PR includes the following changes:

* allow multiple DTLS by using `TimerMilliContext`
* move CoAP::mSocket into Dtls
* move ThreadNetif::mDtls into CoapSecure
* remove unnecessary getters of `Dtls` and `CoapSecure`

### CODE REVIEW

File: src/core/meshcop/dtls.hpp [hunk 1]
```diff
@@ -108,12 +111,22 @@ class Dtls : public InstanceLocator
      * This function pointer is called when data is received from the DTLS session.
      *
      * @param[in]  aContext  A pointer to application-specific context.
-     * @param[in]  aBuf      A pointer to the received data buffer.
+     * @param[in]  aMessage  A pointer to the received data buffer.
```
Comment (jwhui) [inline src/core/meshcop/dtls.hpp]: Is this change intentional?

File: src/core/meshcop/dtls.hpp [hunk 2]
```diff
@@ -125,27 +138,93 @@ class Dtls : public InstanceLocator
      */
     typedef otError (*SendHandler)(void *aContext, const uint8_t *aBuf, uint16_t aLength, uint8_t aMessageSubType);
 
+    /**
+     * This method opens the DTLS socket.
+     *
+     * @param[in]  aHandler     A pointer to a function that is called when receiving DTLS messages.
```
Comment (jwhui) [inline src/core/meshcop/dtls.hpp]: Parameter names not consistent with declaration.

File: src/core/meshcop/dtls.hpp [hunk 3]
```diff
@@ -125,27 +138,93 @@ class Dtls : public InstanceLocator
...
      *
      * For CoAP Secure API do first:
      * Set X509 Pk and Cert for use DTLS mode ECDHE ECDSA with AES 128 CCM 8 or
      * set PreShared Key for use DTLS mode PSK with AES 128 CCM 8.
      *
-     * @param[in]  aClient                 TRUE if operating as a client, FALSE if operating as a server.
-     * @param[in]  aConnectedHandler       A pointer to the connected handler.
-     * @param[in]  aReceiveHandler         A pointer to the receive handler.
-     * @param[in]  aSendHandler            A pointer to the send handler.
      * @param[in]  aContext                A pointer to application-specific context.
```
Comment (jwhui) [inline src/core/meshcop/dtls.hpp]: Fix doxygen.

File: src/core/meshcop/dtls.hpp [hunk 4]
```diff
@@ -125,27 +138,93 @@ class Dtls : public InstanceLocator
...
+    /**
+     * This method binds this DTLS to a UDP port.
+     *
+     * @param[in]  aPort              The port to bind.
+     *
+     * @retval OT_ERROR_NONE        Successfully binded the DTLS service.
+     * @retval OT_ERROR_ALREADY     Already binded.
+     *
+     */
+    otError Start(uint16_t aPort);
```
Comment (jwhui) [inline src/core/meshcop/dtls.hpp]: Should we rename this method to bind?

File: src/core/meshcop/dtls.hpp [hunk 5]
```diff
@@ -125,27 +138,93 @@ class Dtls : public InstanceLocator
...
+     * This method binds this DTLS with a transport callback.
+     *
+     * @param[in]  aCallback  A pointer to a function for sending messages.
+     * @param[in]  aContext   A pointer to arbitrary context information.
+     *
+     * @retval OT_ERROR_NONE        Successfully binded the DTLS service.
+     * @retval OT_ERROR_ALREADY     Already binded.
+     *
+     */
+    otError Start(TransportCallback aCallback, void *aContext);
```
Comment (jwhui) [inline src/core/meshcop/dtls.hpp]: Should we rename this method to bind?

**Comment by bukepo** [comment]
Fixed comments. @jwhui @abtink
